File: elixir/pangram/pangram.exs
defmodule Pangram do
  @doc """
  Determines if a word or sentence is a pangram.
  A pangram is a sentence using every letter of the alphabet at least once.

  Returns a boolean.

    ## Examples

      iex> Pangram.pangram?("the quick brown fox jumps over the lazy dog")
      true

  """

  @spec pangram?(String.t) :: boolean
  def pangram?(""), do: false
  def pangram?(sentence) do
    sentence
    |> String.downcase
    # Remote all non-letter characters.
    |> String.replace(~r/[^a-z]/, "")
    |> String.to_charlist
    |> Enum.sort
    |> pangram_check
  end

  defp pangram_check(text), do: pangram_check(text, ?a - 1)
  # If all letters are gone, and we saw z, success.
  defp pangram_check([], ?z), do: true
  # If all letters are gone before z, fail.
  defp pangram_check([], _), do: false
  defp pangram_check([char | rest], last_char) do
    # If the character is the last one again, or the next one, all good.
    if last_char == char - 1 || last_char == char do
      pangram_check(rest, char)
    else
      false
    end
  end
end

File: ruby/pangram/pangram.rb
# Determine if a sentence is a pangram. A pangram
# is a sentence using every letter of the alphabet at least once.
# The best known English pangram is
# "The quick brown fox jumps over the lazy dog."

# The alphabet used is ASCII, and case insensitive, from 'a' to 'z'
# inclusively.

module BookKeeping
  VERSION = 3
end

class Pangram
  A_POINT = "a".codepoints[0]
  Z_POINT = "z".codepoints[0]

  def self.pangram?(text)
    return false if text == ""

    last_letter =
      text
        .downcase
        .gsub(/[^a-z]/, "") # Remove all non-letter characters.
        .codepoints
        .sort
        .reduce(A_POINT) do |last_char, char|
          break false unless last_char == char - 1 || last_char == char
          char
        end

    last_letter && last_letter == Z_POINT
  end
end
